Plumb dart timeline events to mojo:tracing.

ppiet opened the pull request:
This patch:

 - makes dart traces (coming from the VM and the timeline API)
   available to mojo:tracing
 - adds --complete-timeline parameter, which makes the Dart VM enable all
   categories of traces from the beginning, allowing to do startup tracing

This is based on the equivalent support in dart content handler. In particular,
dart_tracing.{h,cc} are imported with only mechanical changes (includes and
include guard).

ppiet commented:
@chinmaygarde @johnmccutchan - ptal

krisgiesing commented on sky/engine/core/script/dart_init.cc:
```diff
@@ -102,6 +102,10 @@ static const char* kDartStartPausedArgs[]{
 };
 
+static const char* kCompleteTimelineArgs[]{
+    "--complete-timeline",
```
The other flags use underscore conventions here. Should this be `--complete_timeline`?

ppiet replied:
Done.

johnmccutchan commented:
I do not recommend using --complete-timeline for startup tracing.

ppiet commented:
@johnmccutchan what do you recommend instead? Without this option we don't get the timeline traces. (neither from flutter w/ this patch nor from the dart content handler)

ppiet commented:
Can we merge this? As per https://codereview.chromium.org/1893223004/ and offline discussion with @johnmccutchan , we will need a new flag in the dart VM to make this work without `--complete-timeline`. In the meantime, this patch works and is useful (@emembrives has been patching it in locally).

eseidelGoogle commented:
@abarth is probably your correct reviewer, he's OOO until Weds.

johnmccutchan commented:
@ppiet Why can't the Dart traces be fetched over the service protocol directly?

ppiet commented:
@johnmccutchan what is "service protocol"? We need the dart traces to be available to mojo:tracing, so that one can take a unified trace representing the events in dart content handler, flutter content handler, and other mojo apps.

ppiet commented:
@abarth ptal. @johnmccutchan can you explain the point about the service protocol above?

chinmaygarde commented on sky/shell/platform/mojo/dart_tracing.cc:
```diff
@@ -0,0 +1,144 @@
+
+void DartTimelineController::EnableAll() {
+  Dart_GlobalTimelineSetRecordedStreams(DART_TIMELINE_STREAM_DART);
+  Dart_GlobalTimelineSetRecordedStreams(DART_TIMELINE_STREAM_ALL |
```
I believe you can replace both these lines with just `Dart_GlobalTimelineSetRecordedStreams(DART_TIMELINE_STREAM_ALL)`

ppiet replied:
Done.

johnmccutchan commented:
@ppiet The vm service protocol is a JSON RPC protocol for communicating with the VM. It is what Observatory uses.

chinmaygarde commented:
I believe what @johnmccutchan is referring to is connecting to the Dart observatory using a WebSocket connection and make specific requests to start, stop and collect tracing information. Something like https://github.com/flutter/flutter/pull/3393/files.

lgtm as a starting point for getting immediately useful tracing information.

abarth commented on sky/shell/platform/mojo/dart_tracing.cc:
```diff
@@ -0,0 +1,144 @@
+      }
+      curly_braces++;
+    }
+    if (data[i] == '}') {
```
What if these curly braces are inside string literals?

ppiet replied:
Done - switched to use proper JSON parser from base.

abarth commented on sky/engine/public/platform/sky_settings.h:
```diff
@@ -17,6 +17,7 @@ struct SkySettings {
   bool start_paused = false;
   bool enable_dart_checked_mode = false;
   bool trace_startup = false;
+  bool complete_timeline = false;
```
What is "complete timeline" ?  We should have a better name than this.  Why would I ever want an incomplete timeline?

krisgiesing replied:
I believe this nomenclature comes from a Dart VM flag.

abarth replied:
Yeah, but the name doesn't make sense outside that context.  (I'm not even sure what the name means in the context of the Dart VM.)

johnmccutchan replied:
By default the VM timeline uses a ring buffer and old events are dropped.

--complete_timeline is short hand for:

--timeline_streams=all
--timeline_recorder=endless

All timeline categories are enabled and we use the 'endless' recorder which will keep allocating memory for timeline events until malloc fails.

--startup_timeline is short hand for:

--timeline_streams=all
--timeline_recorder=startup

All timeline categories are enabled and we use the 'startup' recorder which will record up to 8k events and then start dropping events.

abarth replied:
That makes sense.  At this layer of the system, we'll probably need a different name given that we'll still use a ring buffer for the `base` trace events.

ppiet replied:
Renamed to "vm-complete-timeline". (BTW, one could argue that that `base` trace events are not "timeline" events, so maybe there was no conflict in the first place)

ppiet commented:
@chinmaygarde @johnmccutchan thanks for clarification regarding the vm service protocol. How would you suggest that we integrate this with mojo:tracing? For tracing mojo apps it is essential that the tracing information is provided to mojo:tracing, so that one can take a unified trace covering all apps that are running.

@abarth thanks, I addressed comments so far, ptal.

johnmccutchan commented:
@ppiet I think that @chinmaygarde explained it. You can fetch the trace information over the service protocol instead of gathering it this way.

ppiet commented:
@johnmccutchan I must be still missing something, because connecting to flutter observatory over websocket (from where?) doesn't seem to be a solution that ends up making mojo calls to the tracing service.

ppiet commented:
Rebased the PR on upstream master. @abarth @chinmaygarde should I squash the commits?

chinmaygarde commented:
Yes please. Thank you.

ppiet commented:
Done.

Hixie commented:
What is the status of this PR?

ppiet commented:
@Hixie good to merge as far as I am concerned.

Hixie commented:
Ok, landing it. Thanks!
